Add load balancer and IP pool docs (#323)
harvester/docs: docs/networking/loadbalancer.md

---
sidebar_position: 4
sidebar_label: Load Balancer
title: "Load Balancer"
keywords:
- Load Balancer
---
_Available as of v1.2.0_

> The Harvester load balancer is a built-in Layer 4 load balancer that distributes incoming traffic across workloads deployed on Harvester virtual machines (VMs) or guest Kubernetes clusters.

Reviewer: ```suggestion
The Harvester load balancer (LB) is a built-in Layer 4 load balancer that distributes incoming traffic across workloads deployed on Harvester virtual machines (VMs) or guest Kubernetes clusters.
```

## VM Load Balancer

### Features

> Harvester VM Load Balancer includes the following features:

Reviewer 2: ```suggestion
Harvester VM Load Balancer supports the following features:
```

> - Address assignment: Get load balancer IP address either from a DHCP server or from a pre-defined IP pool.

Reviewer 2: ```suggestion
- Address assignment: Get a load balancer IP address either from a DHCP server or a pre-defined IP pool.
```

- Protocol support: Supports both TCP and UDP protocols for load balancing.
- Multiple listeners: Multiple listeners can be created to handle incoming traffic on different ports or with different protocols.

> - Label selector: The load balancer matches the backend servers using label selectors. Therefore, you need to configure the corresponding labels for the backend virtual machines that you want to add to the load balancer.

Reviewer 2: ```suggestion
- Label selector: The load balancer uses label selectors to match the backend servers. Therefore, you need to configure the corresponding labels for the backend virtual machines that you want to add to the load balancer.
```

> - Health check: Health checks can be configured to ensure that traffic is only sent to healthy backend instances.

Reviewer 2: ```suggestion
- Health check: Ensures that traffic is only sent to healthy backend instances.
```

 
### Limitations
Harvester VM Load Balancer has the following limitations:

> - Namespace Restriction: The load balancer is limited to using virtual machines located in the same namespace as backend servers. This restriction is in place to facilitate permission management.

Reviewer 2: ```suggestion
- Namespace restriction: The load balancer can only use virtual machines located in the same namespace as the backend servers. This restriction is in place to facilitate permission management.
```

- **IPv4-only:** The LB is only compatible with IPv4 addresses for VMs.

> - **Guest agent installation:** Installing the guest agent on each backend VM  is required to obtain IP addresses.

Reviewer 3: Q: do we support Windows VM in this case?

Author: Let me check it.

Author: We are unable to install the guest agent for the Windows VM. Therefore, Windows VM is not supported.

Author: I'll add a note.

> - IPv4 Only: The load balancer is only compatible with IPv4 addresses for VMs.

Reviewer 2: ```suggestion
- IPv4-only: The load balancer is only compatible with IPv4 addresses for VMs.
```

> - Guest Agent Requirement: To obtain the IP address, the installation of guest agent is required on each backend VM. 

Reviewer 2: ```suggestion
- Guest agent installation: To obtain an IP address, you need to install the guest agent on each backend VM. 
```

- **Connectivity Requirement:** Network connectivity must be established between backend VMs and Harvester hosts. When a VM has multiple IP addresses, the LB will select the first one as the backend address.
- **Access Restriction:** The VM LB address is exposed only within the same network as the Harvester hosts. To access the LB from outside the network, you must provide a route from outside to the LB address.

> ### How to create 

Reviewer 3: consider adding an appropriate screenshot.

> To create a new VM load balancer:

Reviewer: ```suggestion
To create a new Harvester VM load balancer:
```

> To create a new VM load balancer, go to the **Networks > Load Balancer** page and click the **Create** button.

Reviewer 2: ```suggestion
To create a new VM load balancer:
1. Go to the **Networks > Load Balancer** page and click the **Create** button.
```

1. Go to the **Networks > Load Balancer** page and select **Create**.
1. Select the **Namespace** and specify the **Name**.
1. Go to the **Basic** tab to choose the IPAM mode, which can be **DHCP** or **IP Pool**. If you select **IP Pool**, prepare an IP pool first, specify the IP pool name, or choose **auto**. If you choose **auto**, the LB automatically selects an IP pool according to [the IP pool selection policy](/networking/ippool.md/#selection-policy).
1. Go to the **Listeners** tab to add listeners. You must specify the **Port**, **Protocol**, and **Backend Port** for each listener.
1. Go to the **Backend Server Selector** tab to add label selectors. To add the VM to the LB, go to the **Virtual Machine > Instance Labels** tab to add the corresponding labels to the VM.
1. Go to the **Health Check** tab to enable health check and specify the parameters, including the **Port**, **Success Threshold**, **Failure Threshold**, **Interval**, and **Timeout** if the backend service supports health check.
 

> ## Guest Kubernetes cluster load balancer

Reviewer 3: ditto, consider adding an appropriate screentshot

> 1. Select the namespace and specify the name.
> 2. Go to the **Basic** tab to choose the IPAM mode. You can either select **DHCP** or **IP Pool**. If you select **IP Pool**, you need to prepare an IP pool before and specify the IP pool name or choose **auto**. If you choose **auto**, the load balancer will automatically select an IP pool according to the matching rules.
> 3. Go to the **Listeners** tab to add listeners. For each listener, you need to specify the port, protocol and backend port.
> 4. Go to the **Backend Server Selector** tab to add label selectors. If you want to add the VM to the load balancer, you need to go to **Virtual Machine > Instance Labels ** tab to add the corresponding labels to the VM.
> 5. Go to the **Health Check** tab to enable health check and specify the health check parameters including the port, success threshold, failure threshold, interval and timeout if the backend service supports health check.

Reviewer 2: ```suggestion
1. Select the namespace and specify the name.
1. Go to the **Basic** tab to choose the IPAM mode, which can be **DHCP** or **IP Pool**. If you select **IP Pool**, you need to prepare an IP pool first and specify the IP pool name or choose **auto**. If you choose **auto**, the load balancer will automatically select an IP pool according to the matching rules.
1. Go to the **Listeners** tab to add listeners. For each listener, you need to specify the port, protocol, and backend port.
1. Go to the **Backend Server Selector** tab to add label selectors. If you want to add the VM to the load balancer, go to the **Virtual Machine > Instance Labels ** tab to add the corresponding labels to the VM.
1. Go to the **Health Check** tab to enable health check and specify the health check parameters including the port, success threshold, failure threshold, interval, and timeout if the backend service supports health check.
```

> ## Guest Kubernetes Cluster Load Balancer

Reviewer 2: ```suggestion
## Guest Kubernetes cluster Load Balancer
```

> In conjunction with Harvester Cloud Provider, Harvester Load Balancer provides load balancing for load balancer type services in the guest cluster.

Reviewer 2: ```suggestion
In conjunction with Harvester Cloud Provider, Harvester Load Balancer provides load balancing for load balancer services in the guest cluster.
```

In conjunction with Harvester Cloud Provider, the Harvester load balancer provides load balancing for LB services in the guest cluster.

When you create, update, or delete a LB service on a guest cluster with Harvester Cloud Provider, the Harvester Cloud Provider will create a Harvester LB automatically.

> Refer to [Harvester Cloud Provider](/rancher/cloud-provider.md) for more details.

Reviewer: ```suggestion
For more details, refer to [Harvester Cloud Provider](/rancher/cloud-provider.md).
```

> When you create a load balancer type service on a guest cluster with Harvester Cloud Provider, the Harvester Cloud Provider will create a Harvester Load Balancer automatically. The same automatic process applies to updates and deletions.

Reviewer 2: ```suggestion
When you create a load balancer service on a guest cluster with Harvester Cloud Provider, the Harvester Cloud Provider will create a Harvester Load Balancer automatically. The same automation process also applies to updates and deletions.
```

---
sidebar_position: 4
sidebar_label: Load Balancer
title: "Load Balancer"
keywords:
- Load Balancer
---
_Available as of v1.2.0_

> Harvester load balancer is a built-in Layer 4 load balancer that can be used to distribute incoming traffic across workloads deployed on Harvester virtual machines (VMs) or guest Kubernetes clusters.

Reviewer: ```suggestion
The Harvester VM load balancer is a built-in Layer 4 load balancer that distributes incoming traffic across workloads deployed on Harvester virtual machines (VMs) or guest Kubernetes clusters.
```

Author: Harvester VM load balancer and Harvester load balancer are different.

Reviewer: > Harvester VM load balancer and Harvester load balancer are different.

@yaocw2020 thanks for clarifying

## VM load balancer

### Features
Harvester VM load balancer supports the following features:

> - Address assignment: Get a load balancer IP address either from a DHCP server or a pre-defined IP pool.
> - Protocol support: Supports both TCP and UDP protocols for load balancing.
> - Multiple listeners: Multiple listeners can be created to handle incoming traffic on different ports or with different protocols.
> - Label selector: The load balancer uses label selectors to match the backend servers. Therefore, you need to configure the corresponding labels for the backend virtual machines that you want to add to the load balancer.
> - Health check: Ensures that traffic is only sent to healthy backend instances.

Reviewer: ```suggestion
- **Address assignment:** Get the LB IP address from a DHCP server or a pre-defined IP pool.
- **Protocol support:** Supports both TCP and UDP protocols for load balancing.
- **Multiple listeners:** Create multiple listeners to handle incoming traffic on different ports or with other protocols.
- **Label selector:** The LB uses label selectors to match the backend servers. Therefore, you must configure the corresponding labels for the backend VMs you want to add to the LB.
- **Health check:** Only send traffic to healthy backend instances.
```

 
### Limitations
Harvester VM load balancer has the following limitations:

> - Namespace restriction: The load balancer can only use virtual machines located in the same namespace as the backend servers. This restriction is in place to facilitate permission management.
> - IPv4-only: The load balancer is only compatible with IPv4 addresses for VMs.
> - Guest agent installation: To obtain IP addresses, the installation of guest agent is required on each backend VM. 
> - Connectivity Requirement: Network connectivity must be established between backend VMs and Harvester hosts. In the case where a VM has multiple IP addresses, the load balancer will select the first one as the backend address. 
> - Access Restriction: The VM load balancer address is exposed only within the same network as the Harvester hosts. If you wish to access the load balancer from outside the network, you must provide a route from outside to the load balancer address.

Reviewer: ```suggestion
- **Namespace restriction:** This restriction is in place to facilitate permission management and ensures the LB only uses VMs in the same namespace as the backend servers.
- **IPv4-only:** The LB is only compatible with IPv4 addresses for VMs.
- **Guest agent installation:** Installing the guest agent on each backend VM  is required to obtain IP addresses.
- **Connectivity Requirement:** Network connectivity must be established between backend VMs and Harvester hosts. When a VM has multiple IP addresses, the LB will select the first one as the backend address.
- **Access Restriction:** The VM LB address is exposed only within the same network as the Harvester hosts. If you wish to access the LB from outside the network, you must provide a route from outside to the LB address.
```

### How to create 
To create a new VM load balancer:

> 1. Go to the **Networks > Load Balancer** page and click the **Create** button.
> 2. Select the namespace and specify the name.
> 3. Go to the **Basic** tab to choose the IPAM mode, which can be **DHCP** or **IP Pool**. If you select **IP Pool**, you need to prepare an IP pool first and specify the IP pool name or choose **auto**. If you choose **auto**, the load balancer will automatically select an IP pool according to the matching rules.
> 4. Go to the **Listeners** tab to add listeners. For each listener, you need to specify the port, protocol, and backend port.
> 5. Go to the **Backend Server Selector** tab to add label selectors. If you want to add the VM to the load balancer, go to the **Virtual Machine > Instance Labels ** tab to add the corresponding labels to the VM.
> 6. Go to the **Health Check** tab to enable health check and specify the health check parameters including the port, success threshold, failure threshold, interval, and timeout if the backend service supports health check.

Reviewer: ```suggestion
1. Go to the **Networks > Load Balancer** page and select **Create**.
1. Select the **Namespace** and specify the **Name**.
1. Go to the **Basic** tab to choose the IPAM mode, which can be **DHCP** or **IP Pool**. If you select **IP Pool**, you must prepare an IP pool first, specify the IP pool name, or choose **auto**. If you choose **auto**, the LB will automatically select an IP pool according to the matching rules.
1. Go to the **Listeners** tab to add listeners. You must specify the **Port**, **Protocol**, and **Backend Port** for each listener.
1. Go to the **Backend Server Selector** tab to add label selectors. If you want to add the VM to the LB, go to the **Virtual Machine > Instance Labels** tab to add the corresponding labels to the VM.
1. Go to the **Health Check** tab to enable health check and specify the parameters, including the **Port**, **Success Threshold**, **Failure Threshold**, **Interval**, and **Timeout** if the backend service supports health check.
```

 
## Guest Kubernetes cluster load balancer

> In conjunction with Harvester Cloud Provider, Harvester load balancer provides load balancing for load balancer services in the guest cluster.
> When you create a load balancer service on a guest cluster with Harvester Cloud Provider, the Harvester Cloud Provider will create a Harvester load balancer automatically. The same automation process also applies to updates and deletions.
> Refer to [Harvester Cloud Provider](/rancher/cloud-provider.md) for more details.

Reviewer: ```suggestion
In conjunction with Harvester Cloud Provider, the Harvester load balancer provides load balancing for LB services in the guest cluster.

When you create, update, or delete a LB service on a guest cluster with Harvester Cloud Provider, the Harvester Cloud Provider will create a Harvester LB automatically.

Refer to [Harvester Cloud Provider](/rancher/cloud-provider.md) for more details.
```

> 
> 
> 

Reviewer: ```suggestion
```
